core/src/main/java/org/apache/calcite/rel/core/JoinInfo.java

```diff
@@ -59,23 +61,21 @@ public static JoinInfo of(RelNode left, RelNode right, RexNode condition) {
     RexNode remaining =
         RelOptUtil.splitJoinCondition(left, right, condition, leftKeys,
             rightKeys, filterNulls);
-    if (remaining.isAlwaysTrue()) {
-      return new EquiJoinInfo(ImmutableIntList.copyOf(leftKeys),
-          ImmutableIntList.copyOf(rightKeys));
-    } else {
-      return new NonEquiJoinInfo(ImmutableIntList.copyOf(leftKeys),
-          ImmutableIntList.copyOf(rightKeys), remaining);
-    }
+    return new JoinInfo(ImmutableIntList.copyOf(leftKeys),
+        ImmutableIntList.copyOf(rightKeys), remaining);
   }
 
   /** Creates an equi-join. */
   public static JoinInfo of(ImmutableIntList leftKeys,
       ImmutableIntList rightKeys) {
-    return new EquiJoinInfo(leftKeys, rightKeys);
+    return new JoinInfo(leftKeys, rightKeys, null);
   }
 
   /** Returns whether this is an equi-join. */
-  public abstract boolean isEqui();
+  public boolean isEqui() {
+    return (nonEqui == null || nonEqui.isAlwaysTrue())
+        && leftKeys.size() == rightKeys.size();
+  }
```

Comment:
`leftKeys.size()` should always equals with `rightKeys.size()`

Reply:
Correct, this is sanity check.